### DISCUSSION TITLE
Nested variable substitution

### DISCUSSION THREAD

**Discussion by s-boardman** [community/community#26428] [Actions]
<p>I am using a matrix to run my unit tests in a variety of environments. I have a dependencies file per environment. I want to use the hash of this in the cache restore key so I need to define the file to look for using the ${{ matrix.stack }} variable (which is just a string) like so:</p>
<pre><code class="lang-auto">key: ${{ runner.os }}-conda-stage-${{ matrix.stack }}-${{ hashFiles('env/${{ matrix.stack }}_environment.yml') }}
</code></pre>
<p>The nested expansion of ${{ matrix.stack }} within the hashFiles call doesn’t happen so the key defaults to:</p>
<pre><code class="lang-auto">key: ${{ runner.os }}-conda-stage-${{ matrix.stack }}-
</code></pre>
<p>Therefore the cached environment is always restored even when the environment file is changed.</p>

**Comment by Simran-B** [comment, marked as answer]
<p>Nested expressions are not supported I think?</p>
<p>You can do the following however:</p>
<p><code>key: ${{ format('{0}-conda-stage-{1}-{2}', runner.os, matrix.stack, hashFiles(format('env/{0}_environment.yml', matrix.stack))) }}</code></p>
<aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions#format" target="_blank" rel="noopener nofollow ugc">docs.github.com</a>
  </header>
  <article class="onebox-body">
    <img src="" class="thumbnail" width="16" height="16">

<h3><a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions#format" target="_blank" rel="noopener nofollow ugc">Context and expression syntax for GitHub Actions - GitHub Docs</a></h3>



  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by s-boardman** [comment]
<p>Fantastic, that’s worked perfectly. I’d missed the <code>format</code> function! Thanks very much.</p>

**Comment by nrdxp** [comment]
this doesn't work for `fromJSON` if we are building a dynamic matrix from a json and one of the object keys is dynamic: I tried this so far:
`target: ${{ fromJSON(needs.discover.outputs.hits).format('{0}Jobs', inputs,environment).deploy }}`
`target: ${{ fromJSON(needs.discover.outputs.hits).${{ inputs.environment }}Jobs.deploy }}`

For now I have resorted to a setup job to manipulate the json with `jq` and then set that as an output that the deployment reads.

**Reply by pascallj** [reply]
> this doesn't work for `fromJSON` if we are building a dynamic matrix from a json and one of the object keys is dynamic: I tried this so far: `target: ${{ fromJSON(needs.discover.outputs.hits).format('{0}Jobs', inputs,environment).deploy }}` `target: ${{ fromJSON(needs.discover.outputs.hits).${{ inputs.environment }}Jobs.deploy }}`
> 
> For now I have resorted to a setup job to manipulate the json with `jq` and then set that as an output that the deployment reads.

@nrdxp It took me a while to figure this out, but you can use the 'index' notation to accomplish this:

```yaml
target: ${{ fromJSON(needs.discover.outputs.hits)[format('{0}Jobs', input.environment)]['deploy'] }}
```
